src/execution/riscv64/simulator-riscv64.cc

```diff
@@ -694,17 +648,57 @@ void RiscvDebugger::Debug() {
     }
   }
 
-  // Add all the breakpoints back to stop execution and enter the debugger
-  // shell when hit.
-  RedoBreakpoints();
-
 #undef COMMAND_SIZE
 #undef ARG_SIZE
 
 #undef STR
 #undef XSTR
 }
 
+void Simulator::SetBreakpoint(Instruction* location) {
+  for (unsigned i = 0; i < breakpoints_.size(); i++) {
+    if (breakpoints_.at(i).location == location) {
+      PrintF("Existing breakpoint at %p was %s\n",
+             reinterpret_cast<void*>(location),
+             breakpoints_.at(i).enabled ? "disabled" : "enabled");
+      breakpoints_.at(i).enabled = !breakpoints_.at(i).enabled;
+      return;
+    }
+  }
+  Breakpoint new_breakpoint = {location, true};
+  breakpoints_.push_back(new_breakpoint);
+  PrintF("Set a breakpoint at %p\n",
+         reinterpret_cast<void*>(location));
+}
+
+void Simulator::ListBreakpoints() {
+  PrintF("Breakpoints:\n");
+  for (unsigned i = 0; i < breakpoints_.size(); i++) {
+    PrintF("%p  : %s\n",
+           reinterpret_cast<void*>(breakpoints_.at(i).location),
+           breakpoints_.at(i).enabled ? "enabled" : "disabled");
+  }
+}
+
+void Simulator::CheckBreakpoints() {
+  bool hit_a_breakpoint = false;
+  Instruction* pc_ = reinterpret_cast<Instruction*>(get_pc());
+  for (unsigned i = 0; i < breakpoints_.size(); i++) {
+    if ((breakpoints_.at(i).location == pc_) && breakpoints_.at(i).enabled) {
+      hit_a_breakpoint = true;
+      // Disable this breakpoint.
+      breakpoints_.at(i).enabled = false;
```

Comment:
What is the intention of `CheckBreakpoints()`? Why does it disable the breakpoint once it is detected as hit_a_breakpoint?

Reply:
why not break out of the loop once a breakpoint is detected?

Reply:
Thank you for your instruction and advice.

These codes are copied from arm64 simulator. So I don't really understand the developers' intentions.

Maybe it is a mistake like that arm64 simulator does not support the `--stop_sim_at <n>` command when started.

I will extend class `Breakpoint` to support both tbreak and regular break.

> why not break out of the loop once a breakpoint is detected?

In my opinion, we need the loop to execute the left instructions. If we break out of the loop, we have to enter it eventually. 

But it would be a pain to enter the loop again. Such like mutual recursion if we enter `Debug()`, and then `Execute()`, and 'Debug()' again. Maybe there is some way better that I have not realized.

So, the current behavior would be more simple because we just stop at the break point, then enter `Debug()`, finally be back to loop and execute the left instructions.

Reply:
Did I miss something here? `CheckBreakpoints()` does not execute any instruction, it only checks if PC matches a breakpoint, if so enter the debugging.

The loop below is just trying to find a breakpoint matching the PC, it does not execute any instruction. So once one breakpoint is found, we can exit from the loop (i.e., I suggest to insert `break` as shown below). 

Note that the action after `hit_a_breakpoint == true` is outside that loop.

```
for (unsigned i = 0; i < breakpoints_.size(); i++) {
    if ((breakpoints_.at(i).location == pc_) && breakpoints_.at(i).enabled) {
      hit_a_breakpoint = true;
      if (breakpoints_.at(i).temporary) {
        // Disable a temporary breakpoint.
        is_tbreak = true;
        breakpoints_.at(i).enabled = false;
        break; <== add a break to exit the loop once a breakpoint it found
      }
    }
  }

// the following is outside the previous loop
if (hit_a_breakpoint) {
    PrintF("Hit %sa breakpoint at %p.\n",
           is_tbreak ? "and disabled " : "",
           reinterpret_cast<void*>(pc_));
    RiscvDebugger dbg(this);
    dbg.Debug();
  }
```

Reply:
I am sorry to have confused you. I got it wrong.

It is the while loop in `Execute()` which executes instructions. So sorry for my misunderstanding.

As for the for loop in `CheckBreakpoints()`, I agree with you.

Thanks for you help and patience.